doc/user/content/sql/subscribe.md

```diff
@@ -288,6 +296,71 @@ If your row has a unique column key, it is possible to map the update to its cor
 
 In the example above, `Column 1` acts as the column key that uniquely identifies the origin row the update refers to; in case this was unknown, hashing the values from `Column 1` to `Column N` would identify the origin row.
 
+### Alternate output formats
+
+{{< alpha />}}
+
+#### `ENVELOPE UPSERT`
+
+Takes a list of columns to interpret as a key (given by the `KEY` option) and within each distinct timestamp interprets the updates as a series of upserts. The output columns are reordered as `(mz_timestamp, mz_state, k1, k2, .., v1, v2, ..)`. There is no `mz_diff` column. The values `v1, v2, ..` are set the last value if there's an update or an insert and are all set to `NULL` if the only updates to the key in that timestamp were deletions. `mz_state` is either "delete" or "upsert" to distinguish the two cases; when there are no values or the values can all be NULL this column is the only way to distinguish a deletion from an upsert. If `PROGRESS` is set we also return the `mz_progressed` column as usual before `mz_state` and each progress message nulls out all the key and value columns.
+
+Using `ENVELOPE UPSERT` when there is more than one live value per key can be confusing. The upserts generated have no way to express which value was deleted for a given key.
+
+Example:
+
+```nofmt
+> CREATE TABLE kv_store(key int, value int, UNIQUE(key))
+> INSERT INTO kv_store VALUES (1, 2), (2, 4)
+> SUBSCRIBE kv_store ENVELOPE UPSERT (KEY (key))
+mz_timestamp | mz_state | key  | value
+-------------|----------|------|--------
+100          | upsert   | 1    | 2
+100          | upsert   | 2    | 4
+
+...
+-- at time 200, update key=1's value to 10
+...
+
+200          | upsert   | 1    | 10
+
+
+...
+-- at time 300, add a new row with key=3, value=6
+...
+
+300          | upsert   | 3    | 6
+
+...
+-- at time 400, delete all rows
+...
+
+400          | delete   | 1    | NULL
+400          | delete   | 2    | NULL
+400          | delete   | 3    | NULL
+```
+
+#### `WITHIN TIMESTAMP ORDER BY`
+
+`WITHIN TIMESTAMP ORDER BY` takes a `ORDER BY` expression and sorts the returned data within each distinct timestamp.
+This `ORDER BY` can take any column in the underlying object or query in addition to the `mz_diff` column.
+A common use case of recovering upserts to key-value data would be sort by all the keys and `mz_diff`.
+Since `SUBSCRIBE` guarantees that there won't be multiple updates for the same row (the batch is already consolidated), this will produce data that looks like `(k, v1, -1), (k, v2, +1)` which are easier to handle.
+If `PROGRESS` is set, progress messages are unaffected.
```

Comment:
This seems to be the case for both of the new modes, and it might be clearer to remove the sentences about PROGRESS.

Reply:
removed the sentences about progress